### ISSUE TITLE
Show incoming signal connections on editor's Signals panel and as a Node icon on the tree

### ISSUE THREAD

**Issue by MikeFP** [godotengine/godot-proposals#5717]
### Describe the project you are working on

An addon that exposes custom utility Nodes

### Describe the problem or limitation you are having in your project

It's difficult to know exactly which signals are connected to a given Node in the Editor's tree.

If a tool-enabled Node automatically adds default connections when added to the tree, the user may face these problems:

- The user doesn't know how the Node works, so they may not even notice a new connection was made. Even though the Emitting Node displays the Signal icon on the tree, it's not noticeable if the Emitting Node already had signal connections previously.
- The user wants to change the default behavior of the Node, and make new signal connections on the Editor, but they forget to remove the default connections, because the Node does not display any hint about the incoming signal connections.

Also, slightly related, the method that returns those connections `Object.get_incoming_connections()` could have a better name, since some other methods related to signal connections explicitly specify "signal" in their name (`get_signal_list()`, `get_signal_connections()`). For that reason, it took me a long time to find it, since I didn't think about searching for "connection" instead of "signal", but that might be because I was not used to the engine's jargons. Also, the documentation could reference it better, since googling "godot incoming signal connections" returns nothing about it. I had to actually open the Object class docs and Ctrl + F there. That might require a different proposal though.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add an icon to Nodes on the editor tree that have incoming signal connections, and make it visually different from the existing outgoing signal connections icon, so users notice if a new incoming connection was made automatically by a tool script.

Show the incoming connections in the Signals panel, and make them visually different from the outgoing connections, so the user can remove incoming connections without having to look for the Emitting Node.

Make it easier for an user to find `get_incoming_connections()` in the docs, so the user can make use of the method.

Optionally, if breaking compatibility is not a problem, rename `get_incoming_connections()` to include "signal", so users can find the method just by skimming through related methods via code completion on their IDE.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For the "has incoming connections" Node icon, it could be the same icon as the current "has outgoing connections", but with a different color; or be the same icon but with a down arrow. Nodes that have both incoming and outgoing signals could have a third neutral color or icon, although using 3 different colors for the same icon would not be clear at all for the user. So I'm more inclined on using different icons only. An up arrow would be added to the current icon, indicating the Node has outgoing connections. A second icon would be similar, but with a down arrow instead. The third icon would be used if the Node has both kinds of connections, like so:

![tree](https://user-images.githubusercontent.com/9321233/199763428-f4a44e59-3516-4f57-af58-660170c77a9e.png)

An alternative would be to invert the current icon upside down, but then there would be no way to differentiate between Nodes that have just outgoing connections, and both outgoing and incoming connections, unless it looked something like this, which I think is less clear:

![tree_inverted](https://user-images.githubusercontent.com/9321233/199767780-caff020e-c2cf-4228-9317-20f8297c11b9.png)

For the Signals panel, the incoming connections would be grouped by connected method, unlike outgoing connections, which are grouped by emitting signal. Since that would make the UI too crowded, incoming connections would be shown in a separate Tab. The connection icons would be the same, just swapped. Here is an example:

![panel_group](https://user-images.githubusercontent.com/9321233/199778264-0399b77d-3db7-40ef-ad61-27f81242349d.png)

The user could remove incoming connections through the context menu, by right-clicking a signal (the one with the red icon). To facilitate grouping incoming connections by method, the UI could use a new method `get_method_connection_list(method)`, similar to `get_signal_connection_list(signal)`, but to list all incoming connections to a given method, instead of all outgoing connections from the given signal.

For the docs, I think citing the existence and purpose of `get_incoming_connections()` on both `get_signal_list()` and  `get_signal_connection_list()` would be enough:

```
# Returns the list of signals as an [Array] of dictionaries.
# 
# To get the connections between a signal and other objects, use [get_signal_connection_list()].
# To get the connections from external signals to this object, use [get_incoming_connections()].
func get_signal_list( ) -> Array:
    pass

# Returns an [Array] of connections from the given signal to other objects.
# 
# To get the connections from external signals to this object, use [get_incoming_connections()].
func get_signal_connection_list(signal: String) -> Array:
    pass
```

If a rename is desired, I suggest `get_incoming_signal_connections()`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The proposed solutions could be implemented as an EditorPlugin, except for the docs refactor and method rename.

### Is there a reason why this should be core and not an add-on in the asset library?

Editing signal connections is a fundamental engine operation, so it makes more sense to put it in core.

**Comment by dugramen** [comment]
I was just about to propose this, but slightly differently. 
Instead of a 2nd tab for incoming signals, I think it should just be another item above the current list called "Incoming Connections". You expand that item to see the incoming signals, since there will only ever be a few in there. 

And then instead of rotating the signal icon, I think it should be color coded. Green for outgoing signals, Red for incoming signals, and half Red & half Green for both. That way it matches the colors of the icons in the list

**Comment by MikeFP** [comment]
> I was just about to propose this, but slightly differently. Instead of a 2nd tab for incoming signals, I think it should just be another item above the current list called "Incoming Connections". You expand that item to see the incoming signals, since there will only ever be a few in there.

Yeah, considering there will only ever be a few incoming connections, it wouldn't be as crowded as I thought. It might work.

> And then instead of rotating the signal icon, I think it should be color coded. Green for outgoing signals, Red for incoming signals, and half Red & half Green for both. That way it matches the colors of the icons in the list

That's something to be tested. I also suggested that as an alternative, but using different colors with the same icon is a little less readable, specially for colorblind people (me). For them to be distinct enough, they would have to be very saturated (basically, the same colors of the list, as you pointed out), but then, they would stand out too much in relation to other icons in the tree, which are less saturated overall, so they don't draw as much attention.

> Also coupled with #10129, instead of highlighting the whole row in the tree, you could just highlight the icon

Just looked over the proposal, it looks great. I think highlighting the whole row is always more readable. Maybe the icon could be highlighted too somehow, but to differentiate whether you are inspecting an outgoing signal or an incoming connection.
